src/main/java/codeu/controller/BotServlet.java

```diff
@@ -0,0 +1,149 @@
+        userStore = UserStore.getInstance();
+        conversationStore = ConversationStore.getInstance();
+    }
+
+    /**
+     * Sets the MessageStore used by this servlet. This function provides a common setup method for
+     * use by the test framework or the servlet's init() function.
+     */
+    void setMessageStore(MessageStore messageStore) {
+        this.messageStore = messageStore;
+    }
+    /**
+     * This function fires when a user navigates to the bot chat page. It gets the conversation title "bot-conversation" from
+     * the URL, finds the corresponding Conversation, and fetches the messages in that Conversation.
+     * It then forwards to bot chat.jsp for rendering.
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws IOException, ServletException {
+        String username = (String) request.getSession().getAttribute("user");
+        if (username == null) {
+            //redirect to login
+            response.sendRedirect("/login");
+            return;
+        }
+        User user = userStore.getUser(username);
+        if (user == null) {
+            // couldn't find author, redirect to loging Page
+            response.sendRedirect("/login");
+            return;
+        }
+
+        Conversation botConversation = conversationStore.getBotConversation(user.getId());
+        if (botConversation == null){
+             botConversation = new Conversation(UUID.randomUUID(), user.getId(), DEFULT_BOT_CONVERSATION_TITLE+username, Instant.now());
+            conversationStore.addConversation(botConversation);
+        }
+        List<Message> messages = messageStore.getMessagesInConversation(botConversation.getId());
+        request.setAttribute("botmessage", messages);
+        request.getRequestDispatcher("/WEB-INF/view/bot.jsp").forward(request, response);
+    }
+
+    /**
+     * This function fires when a user submits the form on the bot chat page. It gets the logged-in
+     * username from the session,  and the chat message from the
+     * submitted form data. It creates a new Message from that data, adds it to the model, and then
+     * redirects back to the chat page.
+     */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response)
+            throws IOException, ServletException {
+
+        String username = (String) request.getSession().getAttribute("user");
+        if (username == null) {
+            response.sendRedirect("/login");
+            return;
+        }
+        User user = userStore.getUser(username);
+        if (user == null) {
+            response.sendRedirect("/login");
+            return;
+        }
+
+        String userInput = request.getParameter("botmessage");
+
+        Conversation botConversation = conversationStore.getConversationWithTitle(DEFULT_BOT_CONVERSATION_TITLE+username);
+        if (botConversation == null){
```

Comment:
In theory, you don't have to validate if the post is null in the post method. Nevertheless one can't always be so sure.

Reply:
got it.